versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractWriteUpdateCommits.java

```diff
@@ -0,0 +1,122 @@
+
+public abstract class AbstractWriteUpdateCommits {
+
+  private final DatabaseAdapter databaseAdapter;
+  private final boolean commitWritesValidated;
+
+  protected AbstractWriteUpdateCommits(
+      DatabaseAdapter databaseAdapter, boolean commitWritesValidated) {
+    this.databaseAdapter = databaseAdapter;
+    this.commitWritesValidated = commitWritesValidated;
+  }
+
+  Hash intAsHash(int i) {
+    return Hash.of(String.format("%08x", i));
+  }
+
+  private CommitLogEntry helloCommit(int i) {
+    return CommitLogEntry.of(
+        i,
+        intAsHash(i),
+        i,
+        singletonList(intAsHash(i - 1)),
+        ByteString.copyFromUtf8("hello " + i),
+        emptyList(),
+        emptyList(),
+        0,
+        null,
+        emptyList(),
+        emptyList(),
+        emptyList());
+  }
+
+  @Test
+  void writeMultipleCommits() throws Exception {
+    List<Hash> hashes =
+        IntStream.rangeClosed(1, 10).mapToObj(this::intAsHash).collect(Collectors.toList());
+    List<CommitLogEntry> commits =
+        hashes.stream()
+            .mapToInt(h -> Integer.parseInt(h.asString(), 16))
+            .mapToObj(this::helloCommit)
+            .collect(Collectors.toList());
+
+    databaseAdapter.writeMultipleCommits(commits);
+
+    try (Stream<CommitLogEntry> retrieved =
+        databaseAdapter.fetchCommitLogEntries(hashes.stream())) {
+      assertThat(retrieved).containsExactlyInAnyOrderElementsOf(commits);
+    }
+
+    List<CommitLogEntry> updatedCommits =
+        commits.stream()
+            .map(
+                commit ->
+                    ImmutableCommitLogEntry.builder()
+                        .from(commit)
+                        .metadata(ByteString.copyFromUtf8("updated " + commit.getHash().asString()))
+                        .build())
+            .collect(Collectors.toList());
+
+    databaseAdapter.updateMultipleCommits(updatedCommits);
+
+    try (Stream<CommitLogEntry> retrieved =
+        databaseAdapter.fetchCommitLogEntries(hashes.stream())) {
+      assertThat(retrieved).containsExactlyInAnyOrderElementsOf(updatedCommits);
+    }
+  }
+
+  @Test
+  void validateWriteUpdateCommitPreconditions() throws Exception {
+    assumeThat(commitWritesValidated).isTrue();
+
+    List<Hash> hashes =
+        IntStream.rangeClosed(1, 10).mapToObj(this::intAsHash).collect(Collectors.toList());
+    List<CommitLogEntry> commits =
+        hashes.stream()
+            .mapToInt(h -> Integer.parseInt(h.asString(), 16))
+            .mapToObj(this::helloCommit)
+            .collect(Collectors.toList());
+
+    databaseAdapter.writeMultipleCommits(commits.subList(0, 2));
+    assertThatThrownBy(() -> databaseAdapter.writeMultipleCommits(commits.subList(0, 2)))
+        .isInstanceOf(ReferenceConflictException.class);
+
+    assertThatThrownBy(() -> databaseAdapter.updateMultipleCommits(commits))
+        .isInstanceOf(ReferenceNotFoundException.class);
```

Comment:
do the first 2 get updated? or does the whole batch fail atomically?

i would assume the latter to be true for the TX adapters for example.
would this be worth testing explicitly?

Reply:
so i guess we dont want to test the TX adapter as strictly as possible here, fine i guess when we dont want users to rely on this exact behavior